Raven.Client.Lightweight/Changes/RemoteChangesClientBase.cs

```diff
@@ -201,83 +213,133 @@ private bool DisposeHeartbeatTimer()
             }
         }
 
-        private Task lastSendTask;
-
-        protected Task Send(string command, string value)
+        async Task Worker()
         {
-            lock (this)
+            while (tokenSource.IsCancellationRequested == false)
             {
-                logger.Info("Sending command {0} - {1} to {2} with id {3}", command, value, url, id);
-                var sendTask = lastSendTask;
-                if (sendTask != null)
+                try
                 {
-                    return sendTask.ContinueWith(_ =>
-                    {
-                        Send(command, value);
-                    });
+                    await syncSource.Task.ConfigureAwait(false);
+                    syncSource = TaskCompletionSourceFactory.Create<bool>();
                 }
-
-                try
+                catch (TaskCanceledException)
                 {
-                    var sendUrl = url + "/changes/config?id=" + id + "&command=" + command;
-                    if (string.IsNullOrEmpty(value) == false)
-                        sendUrl += "&value=" + Uri.EscapeUriString(value);
+                    // Swallowing the task cancellation in case we are stopping work.
```

Comment:
Shouldn't you exit here and `SetException` for all tasks?

Reply:
No why? I could add a break and then we don't try to empty twice. Judging by the disconnect in dispose we always want to complete all enqueued things right?